lib/resty/rediscluster.lua

```diff
@@ -250,7 +260,10 @@ function _M.fetch_slots(self)
     local _, errors = try_hosts_slots(self, serv_list_combined)
     if errors then
         local err = "failed to fetch slots: " .. table.concat(errors, ";")
-        ngx.log(ngx.ERR, err)
+        ngx_log(ngx_err, err)
+
+        -- clear cached Redis IP and slots info when Redis is down
+        clear_slot_cache(self)
```

Comment:
What if a connection is temporarily unavailable? Or if only part of the nodes are not reachable? Seems that both cases would cause re-discovery.

Reply:
If the connection is temporarily unavailable and then back again online, the cache would be automatically overwritten at here: https://github.com/Kong/resty-redis-cluster/blob/master/lib/resty/rediscluster.lua#L183.

So we are safe to clear this cache under connection failure.

1. If IPs are all changed, then clearing the cache can reduce the time of recovery taken.
2. If IPs are temporarily down, then clearing is safe as it would be overwritten.

Reply:
> 2. If IPs are temporarily down, then clearing is safe as it would be overwritten.

I don't think so. They would only be overwritten once a connection is again established. And then it gets all the active nodes from the cluster again. 

But it would be unsafe in this scenario;

1. cluster starts 3 nodes
2. client is configured with the 3 nodes
3. client connects and requests all nodes from the cluster; gets the same 3 nodes as in the config
4. over time the 3 nodes get recycled, and after a while all 3 original nodes are gone, replaced with new nodes on different addresses
5. because the client occasionally queries the cluster for nodes, the client has a node list containing the 3 new nodes
6. Now the connection fails
7. the client will try to connect to; the 3 last cluster members + the 3 nodes from the config
8. reconnect fails
9. repeat step 7+8 for a while
10. Connection gets restored
11. the client will try to connect to; the 3 last cluster members + the 3 nodes from the config
12. success, cluster list is refreshed, all is well again

This PR would introduce a step `8a:` "drop the list with the last-known cluster members, from here on only use the 3 nodes in the config to reconnect".

Which means that if the cluster returns in its last state (3 nodes that were not in the original config), it will be unable to recover

Reply:
I get what you're saying, but it goes against the design of the Redis Cluster. This is all about high availability and resilience. The more options we keep open to reconnect to the cluster, the better the resilience. So we should not drop the cache.

Reply:
1. The cache only helps HA in a short period: once the worker process is kill and new worker does not have the cache.
2. Drop the cache can improve performance in most cases. IMO, the gains outweigh the loss.

Reply:
Not sure why anyone would kill the worker?

Anyway, we're at a tie here and don't agree. So we need more opinions. @StarlightIbuki @jschmid1 Can you break the tie? (both of you popped up on the potential reviewers list)

Reply:
@Tieske , ha, we should use `reload` instead of `kill`. It is not unusual that we reload Kong, so new workers does not have cache, and failed to connect.

Reply:
@Tieske I've talked offline with @outsinre. This PR should be for cases where we configure with hostname (not IP), and if we only enable this feature (drop cache) when DB is configured with a hostname, the scenario you described should be fine.

Reply:
@Tieske @StarlightIbuki , how about we **only** clear the cache if Kong is configured to use hostname? So when Kong is configured to use IP, we do not touch the cache.

Reply:
I'd rather say flip the order, so first try and reconnect to the configured nodes, and only if those all fail, try and reconnect to the last known nodes from the cache. That would have improved efficiency, but would still try all known nodes for maximum reliability.

Reply:
> I'd rather say flip the order, so first try and reconnect to the configured nodes, and only if those all fail, try and reconnect to the last known nodes from the cache. That would have improved efficiency, but would still try all known nodes for maximum reliability.

That is what the code is doing now. Each configured node and cached node is tried 3 times. If all nodes fail after 3 retries, we clear the cache.

Reply:
I agree with @Tieske. The solution of reordering is more reliable and simple.

Reply:
> I agree with @Tieske. The solution of reordering is more reliable and simple.

That is good. The only downside is it takes longer for Kong to report the failure. Say there are 3 configured IPs and 10 cached IPs, the time it takes is around `(3+10)*2 = 26` seconds.

Reply:
@Tieske agree. I have updated the code. Please help review and then I can work on https://github.com/Kong/resty-redis-cluster/pull/18.